google/cloud/storage/async_writer_test.cc

```diff
@@ -0,0 +1,156 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/async_writer.h"
+#include "google/cloud/storage/mocks/mock_async_writer_connection.h"
+#include "google/cloud/storage/testing/canonical_errors.h"
+#include "google/cloud/testing_util/status_matchers.h"
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+namespace storage_experimental {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+using ::google::cloud::storage::testing::canonical_errors::PermanentError;
+using ::google::cloud::storage_experimental::WritePayload;
+using ::google::cloud::storage_mocks::MockAsyncWriterConnection;
+using ::google::cloud::testing_util::StatusIs;
+using ::testing::ElementsAre;
+using ::testing::IsEmpty;
+using ::testing::ResultOf;
+using ::testing::Return;
+
+template <typename Matcher>
+auto WritePayloadContents(Matcher&& m) {
+  return ResultOf(
+      "WritePayload contents are",
+      [](WritePayload const& p) { return p.payload(); },
+      std::forward<Matcher>(m));
+}
+
+TEST(AsyncWriterTest, Basic) {
+  auto mock = std::make_unique<MockAsyncWriterConnection>();
+  ::testing::InSequence sequence;
+  EXPECT_CALL(*mock, UploadId).WillOnce(Return("test-upload-id"));
+  EXPECT_CALL(*mock, PersistedState).WillOnce(Return(16384));
+  EXPECT_CALL(*mock, Write(WritePayloadContents(ElementsAre("aaa"))))
+      .WillOnce([] { return make_ready_future(Status{}); });
+  EXPECT_CALL(*mock, Write(WritePayloadContents(ElementsAre("bbb"))))
+      .WillOnce([] { return make_ready_future(Status{}); });
+  EXPECT_CALL(*mock, Finalize(WritePayloadContents(ElementsAre("ccc"))))
+      .WillOnce([] {
+        return make_ready_future(make_status_or(storage::ObjectMetadata()));
+      });
+
+  auto token = storage_internal::MakeAsyncToken(mock.get());
+  AsyncWriter writer(std::move(mock));
+  EXPECT_EQ(writer.UploadId(), "test-upload-id");
+  EXPECT_EQ(absl::get<std::int64_t>(writer.PersistedState()), 16384);
+  token = writer.Write(std::move(token), WritePayload{std::string("aaa")})
+              .get()
+              .value();
```

Comment:
shouldn't we `ASSERT_STATUS_OK()` before using the value?

Reply:
I think we are fine: `.value()` throws and stops the test too.